Use std::optional for IDBResourceIdentifier in IDBRequestData and TransactionOperation

szewai opened the pull request:
#### 2f9d01af63cb2f223c20b9bb2a87c99539da2997
<pre>
Use std::optional for IDBResourceIdentifier in IDBRequestData and TransactionOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=244996">https://bugs.webkit.org/show_bug.cgi?id=244996</a>

Reviewed by Chris Dumez.

IDBResourceIdentifier is small enough and members in IDBRequestData and TransactionOperation are not changed after
initialization.

* Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp:
(WebCore::IDBClient::TransactionOperation::TransactionOperation):
* Source/WebCore/Modules/indexeddb/client/TransactionOperation.h:
(WebCore::IDBClient::TransactionOperation::cursorIdentifier const):
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
(WebCore::IDBRequestData::IDBRequestData):
(WebCore::IDBRequestData::isolatedCopy):
(WebCore::IDBRequestData::requestIdentifier const):
(WebCore::IDBRequestData::databaseIdentifier const):
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.h:
(WebCore::IDBRequestData::IDBRequestData): Deleted.
(WebCore::IDBRequestData::databaseIdentifier const): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/254402@main">https://commits.webkit.org/254402@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c12c23d04223da70cab4a1c161819c494d18589

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33511 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19824 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98226 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/154615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92941 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32017 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27614 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81233 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92765 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25426 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75924 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25365 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80294 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/80317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68332 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29808 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29539 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/15339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3084 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32978 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/38281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31667 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/34441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -216,10 +216,10 @@ class WebCore::IDBValue {
+    WebCore::IDBResourceIdentifier m_requestIdentifier;
+    std::optional<WebCore::IDBResourceIdentifier> m_transactionIdentifier;
+    std::optional<WebCore::IDBResourceIdentifier> m_cursorIdentifier;
+        uint64_t m_objectStoreIdentifier;
```
indentation issue.

cdumez commented on Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
```diff
@@ -145,6 +154,14 @@ IndexedDB::IndexRecordType IDBRequestData::indexRecordType() const
+    ASSERT(m_databaseIdentifier);
+    if (!m_databaseIdentifier)
+        return IDBDatabaseIdentifier { };
+    return *m_databaseIdentifier;
```
Seems like a use-case for value_or()

cdumez commented on Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
```diff
@@ -86,6 +97,7 @@ IDBRequestData IDBRequestData::isolatedCopy() const
 void IDBRequestData::isolatedCopy(const IDBRequestData& source, IDBRequestData& destination)
 {
     destination.m_serverConnectionIdentifier = source.m_serverConnectionIdentifier;
+    destination.m_requestIdentifier = source.m_requestIdentifier.isolatedCopy();
```
isolatedCopy() seems unnecessary for an identifier.

cdumez commented on Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
```diff
@@ -94,13 +106,10 @@ void IDBRequestData::isolatedCopy(const IDBRequestData& source, IDBRequestData&
-        destination.m_requestIdentifier = makeUnique<IDBResourceIdentifier>(*source.m_requestIdentifier);
     if (source.m_transactionIdentifier)
-        destination.m_transactionIdentifier = makeUnique<IDBResourceIdentifier>(*source.m_transactionIdentifier);
+        destination.m_transactionIdentifier = source.m_transactionIdentifier->isolatedCopy();
```
isolatedCopy() seems unnecessary.

cdumez commented on Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
```diff
@@ -94,13 +106,10 @@ void IDBRequestData::isolatedCopy(const IDBRequestData& source, IDBRequestData&
+        destination.m_transactionIdentifier = source.m_transactionIdentifier->isolatedCopy();
     if (source.m_cursorIdentifier)
-        destination.m_cursorIdentifier = makeUnique<IDBResourceIdentifier>(*source.m_cursorIdentifier);
+        destination.m_cursorIdentifier = source.m_cursorIdentifier->isolatedCopy();
```
isolatedCopy() seems unnecessary.

cdumez reviewed (approved):
LGTM

szewai committed:
Use std::optional for IDBResourceIdentifier in IDBRequestData and TransactionOperation
https://bugs.webkit.org/show_bug.cgi?id=244996

Reviewed by Chris Dumez.

IDBResourceIdentifier is small enough and members in IDBRequestData and TransactionOperation are not changed after
initialization.

* Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp:
(WebCore::IDBClient::TransactionOperation::TransactionOperation):
* Source/WebCore/Modules/indexeddb/client/TransactionOperation.h:
(WebCore::IDBClient::TransactionOperation::cursorIdentifier const):
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
(WebCore::IDBRequestData::IDBRequestData):
(WebCore::IDBRequestData::isolatedCopy):
(WebCore::IDBRequestData::requestIdentifier const):
(WebCore::IDBRequestData::databaseIdentifier const):
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.h:
(WebCore::IDBRequestData::IDBRequestData): Deleted.
(WebCore::IDBRequestData::databaseIdentifier const): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/254402@main
